lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,409 @@
+// only translates the function itself but no dependencies.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/GPUToROCM/GPUToROCMPass.h"
+
+#include "mlir/Dialect/GPU/GPUDialect.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/Module.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassRegistry.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Target/ROCDLIR.h"
+
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/TargetRegistry.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Transforms/IPO/Internalize.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+using namespace mlir;
+
+namespace {
+
+/// A pass converting tagged kernel functions to HSA Code Object blobs.
+///
+/// If tagged as a kernel module, each contained function is translated to ROCDL
+/// IR, which is then compiled using the llvm AMDGPU backend to generate the GPU
+/// binary code (i.e. the HSACO file). The HSACO binary blob is attached as an
+/// attribute to the function and the function body is erased.
+class GpuKernelToHSACOPass : public ModulePass<GpuKernelToHSACOPass> {
+public:
+  GpuKernelToHSACOPass(rocm::HSACOGeneratorConfig hsacoGeneratorConfig =
+                           rocm::HSACOGeneratorConfig(/*isTestMode*/ true))
+      : config(hsacoGeneratorConfig) {}
+
+  // Run the dialect converter on the module.
+  void runOnModule() override {
+
+    // Nothing to do if this module does not contain the "gpu.kernel_module"
+    // attribute,
+    // which is used to mark the (nested) modules created to house the GPU
+    // kernel functions
+    if (!getModule().getAttrOfType<UnitAttr>(
+            gpu::GPUDialect::getKernelModuleAttrName())) {
+      return;
+    }
+
+    // This is a module containing a GPU kernel function, we have work to do!
+
+    // Make sure the AMDGPU target is initialized.
+    LLVMInitializeAMDGPUTarget();
+    LLVMInitializeAMDGPUTargetInfo();
+    LLVMInitializeAMDGPUTargetMC();
+    LLVMInitializeAMDGPUAsmPrinter();
+
+    auto llvmModule = translateModuleToROCDLIR(getModule());
+    if (!llvmModule) {
+      return signalPassFailure();
+    }
+
+    for (auto function : getModule().getOps<FuncOp>()) {
+      if (!gpu::GPUDialect::isKernel(function)) {
+        continue;
+      }
+
+      if (failed(translateGpuKernelToHSACOAnnotation(*llvmModule, function)))
+        signalPassFailure();
+    }
+
+    // dump the MLIR
```

Comment:
Can we have a more descriptive comment here? Also, why have this in the first place? We have pass manager instrumentation for printing the IR after passes, and environment variables are generally discouraged for this type of thing.

Reply:
this will be removed in the next push